util/encoding/encoding.go

```diff
@@ -370,25 +373,27 @@ func EncodeBytes(b []byte, data []byte) []byte {
 // the decoded []byte are returned.
 func DecodeBytes(b []byte) ([]byte, []byte) {
 	var r []byte
-	copyStart := 0
-	for i, n := 0, len(b)-1; i < n; i++ {
-		v := b[i]
-		if v == escape {
-			v = b[i+1]
-			if v == escapedTerm {
-				if r == nil {
-					return b[i+2:], b[:i]
-				}
-				r = append(r, b[copyStart:i]...)
-				return b[i+2:], r
-			}
-			r = append(r, b[copyStart:i]...)
-			i++
-			if v == escapedNul {
-				r = append(r, 0)
+	for {
+		i := bytes.IndexByte(b, escape)
+		if i == -1 {
+			panic("did not find terminator")
+		}
+		if i+1 > len(b) {
+			panic("malformed escape")
+		}
+		v := b[i+1]
+		if v == escapedTerm {
+			if r == nil {
+				r = b[:i]
+			} else {
+				r = append(r, b[:i]...)
```

Comment:
Where are you seeing the input buffer being modified? It shouldn't be. The above line appends `b[:i]` to `r`. Sometimes we set `r` to `b[:i]`, but only when we can return a portion of the input buffer unmodified as the result.